compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -696,22 +697,45 @@ trait Implicits { self: Typer =>
         }
     }
     def location(preposition: String) = if (where.isEmpty) "" else s" $preposition $where"
+    def userDefinedMessage(annot: Annotation, params: List[String], args: List[Type]): Option[String] =
+      for (Trees.Literal(Constant(raw: String)) <- annot.argument(0)) yield {
+        err.userDefinedErrorString(
+          raw,
+          params,
+          args)
+      }
     arg.tpe match {
       case ambi: AmbiguousImplicits =>
-        msg(s"ambiguous implicit arguments: ${ambi.explanation}${location("of")}")(
-            s"ambiguous implicit arguments of type ${pt.show} found${location("for")}")
-      case _ =>
-        val userDefined =
-          for {
-            notFound <- pt.typeSymbol.getAnnotation(defn.ImplicitNotFoundAnnot)
-            Trees.Literal(Constant(raw: String)) <- notFound.argument(0)
+        val maybeAnnot = ambi.alt1.ref.symbol.getAnnotation(defn.ImplicitAmbiguousAnnot).map(
+          (_, ambi.alt1)
+        ).orElse(
+          ambi.alt2.ref.symbol.getAnnotation(defn.ImplicitAmbiguousAnnot).map(
+            (_, ambi.alt2)
+          )
+        )
```

Comment:
Hm, can this be abstracted with `for { alt <- List(ambi.alt1, ambi.alt2) } .... alt.ref.symbol...`? Or better, in terms of a new `ambi.alts`?

(Except if people feel this is too slow and we're in a performance-critical path).

Actually surprised the error only lists two implicits, apparently since this class' birth (in ea640a32264cb78efbf267d5c2be89e3e99dcccf) only the first two candidate implicits are reported.

Potential bug (not blocking the PR): if you have *three* equally good candidates that are ambiguous, the `@implicitAmbiguous` annotation will be ignored on the third candidate, which is against the specification in http://www.scala-lang.org/api/2.12.3/scala/annotation/implicitAmbiguous.html. If that bug can indeed be triggered (haven't tried, but it's not clear why that shouldn't happen), fixing it seems to require a refactoring of the search algorithm to return all the ambiguous candidates — which is, again, *out of scope* for the PR.

Reply:
Yes I also thought about this. I think the Scalac implementation has the same limitation. I suspect the most realistic fix would probably be in the documentation of `@implicitAmbiguous`.

Reply:
Ah, fair enough — this wasn't a well-known issue around here. Not sure how to specify it, 
so for now I've opened https://github.com/scala/bug/issues/10739?

For the PR, please go ahead!